jaxopt/_src/broyden.py

```diff
@@ -0,0 +1,382 @@
+
+
+def matvec(d_history, c_history, x, indices):
+  # CD^T x
+  # here j is the history dimension and i is the space dimension
+  dtx = jnp.einsum('j..., ... -> j', d_history[indices], x)
+  cdtx = jnp.einsum('j..., j -> ...', c_history[indices], dtx)
+  return cdtx
+
+def inv_jacobian_product_leaf(v: jnp.ndarray,
+                             d_history: jnp.ndarray,
+                             c_history: jnp.ndarray,
+                             gamma: float = 1.0,
+                             start: int= 0):
+  # the computation of the jacobian inverse for
+  # the Broyden method
+  # this is done using the notations of DEQ
+  history_size = len(d_history)
+
+  indices = (start + jnp.arange(history_size)) % history_size
+
+  v = gamma * v + matvec(d_history, c_history, v, indices)
+  return v
+
+
+def inv_jacobian_product(pytree: Any,
+                        d_history: Any,
+                        c_history: Any,
+                        gamma: float = 1.0,
+                        start: int = 0):
+  """Product between an approximate jacobian inverse and a pytree.
+
+  Histories are pytrees of the same structure as `pytree`.
+  Leaves are arrays of shape `(history_size, ...)`, where
+  `...` means the same shape as `pytree`'s leaves.
+
+  The notation follows the scipy one.
+
+  Args:
+    pytree: pytree to multiply with.
+    d_history: pytree with the same structure as `pytree`.
+      Leaves contain v variables, i.e., `(x[k] - x[k-1])^T B / ((g[k] - g[k-1])^T (x[k] - x[k-1])^T B)`.
+    c_history: pytree with the same structure as `pytree`.
+      Leaves contain u variables, i.e., `(x[k] - x[k-1]) - B(g[k] - g[k-1])`.
+    gamma: scalar to use for the initial inverse jacobian approximation,
+      i.e., `gamma * I`.
+    start: starting index in the circular buffer.
+  """
+  fun = partial(inv_jacobian_product_leaf,
+                gamma=gamma,
+                start=start)
+  return tree_map(fun, pytree, d_history, c_history)
+
+def inv_jacobian_rproduct(pytree: Any,
+                        d_history: Any,
+                        c_history: Any,
+                        gamma: float = 1.0,
+                        start: int = 0):
+  return inv_jacobian_product(pytree, c_history, d_history, jnp.conjugate(gamma), start)
+
+
+def init_history(pytree, history_size):
+  fun = lambda leaf: jnp.zeros((history_size,) + leaf.shape, dtype=leaf.dtype)
+  return tree_map(fun, pytree)
+
+
+def update_history(history_pytree, new_pytree, last):
+  fun = lambda history_array, new_value: history_array.at[last].set(new_value)
+  return tree_map(fun, history_pytree, new_pytree)
+
+
+class BroydenState(NamedTuple):
+  """Named tuple containing state information."""
+  iter_num: int
+  value: float
+  stepsize: float
+  error: float
+  d_history: Any
+  c_history: Any
+  gamma: jnp.ndarray
+  aux: Optional[Any] = None
+  failed_linesearch: bool = False
+
+
+@dataclass(eq=False)
+class Broyden(base.IterativeSolver):
+  """Broyden solver.
+
```

Comment:
Could you add here a brief explanation of what are the key differences with LBFGS?

Reply:
sure, wdyt now?

Reply:
LGTM, thanks!